ADIOS plugin

michaelsippel opened the pull request:
This PR adds the plug-in implementation to browse ADIOS-files with HDF-Compass.
Additional dependencies for using this plug-in are: ADIOS (1.9.0+) with numpy wrapper (1.9.1b18+).

**Note:** currently this implementation was only tested with python2 since the unit-tests from HDF-Compass don't work with python3. When these errors are fixed, I will check the support for the ADIOS-plugin too.

CC @ax3l

ax3l committed:
Add ADIOS example Data Set

From official ADIOS repository, file:
  `wrappers/numpy/tests/temp.bp`

ax3l committed:
ADIOS Template: Start with a HDF5 Copy

Commit:
adiosstore & import in load_plugins

Commit:
load_plugins: disable hdf5_rest; add can_handle

Commit:
adios: create root-group with dummy entries

Commit:
split adios keys for groups

Commit:
better adios testfile

Commit:
fix subgroups

Commit:
adios: keyvalue & text

Commit:
adios: create keylist later

Commit:
adios: fix attributes

Commit:
adios unittests

ax3l commented on .travis.yml:
```diff
@@ -63,6 +63,7 @@ script:
   - "python -m unittest hdf_compass.filesystem_model.test"
   - "python -m unittest hdf_compass.hdf5_model.test"
   - "python -m unittest hdf_compass.opendap_model.test"
+  - "python -m unittest hdf_compass.adios_model.test"
```
@michaelsippel below the `conda install -q h5py lxml ...` line please add a new `pip install adios` line so we install the adios numpy wrapper in travis, too :)
(not yet packed in conda)

ax3l replied:
ah I just realized we will need the c lib to be installed in a very recent version, too...

ax3l replied:
I will add some install lines for it, otherwise I can still build a conda recipe

ax3l commented on hdf_compass/compass_viewer/viewer.py:
```diff
@@ -236,11 +236,17 @@ def load_plugins():
         log.warning("Opendap plugin: NOT loaded")
     
-    from hdf_compass import hdf5rest_model    
+#    from hdf_compass import hdf5rest_model    
```
@michaelsippel this looks like it just did not work on your system, but should not be removed in general

Commit:
install adios in travis; uncomment model loading in viewer.py

ax3l committed:
ADIOS: Install & cache c lib

Installs and caches the ADIOS c library from
its current master.

ax3l commented:
@michaelsippel can you please fix the left over travis errors with the example file?
See: https://travis-ci.org/ComputationalRadiationPhysics/hdf-compass/jobs/143174332#L2627

ax3l committed:
ADIOS: Add Copyright & License

Directly dopied from ADIOS mainline `COPYING` file.

michaelsippel commented:
@ax3l My guess is the adios-numpy wrappers wich are installed by pip are outdated. We need to build them in travis from source.

ax3l commented:
Looks like it, can do. (update: done)
Or I ask them if they can release a recent version on pypi. (update: https://github.com/ornladios/ADIOS/issues/75; update `setup.py` requirement after that, too!)

ax3l committed:
ADIOS: Install Python Wrapper from src

Install ADIOS python wrapper from source until next PyPi
release.

ax3l committed:
OSX: Build MPI for ADIOS

ax3l commented:
@jreadey @giumas are you interested in giving us a review (and merge)? :)

jreadey commented:
@ax3l I emailed a @michaelsippel a copy of our contributors agreement, but I haven't heard back yet.  Actually it would be best for you to fill it out as well.  Once we have these on file, we'll be able to merge this this PR and any future ones.

Also, can you point me to some sample files in the ADIOS format?

ax3l commented:
@jreadey thank you, we just sent you the documents back.

We have committed a small adios `.bp` example file with this PR. They come in two flavours: an "all-in-one" file `.bp` and a "distributed" format consisting of a `.bp` file with meta information and a directory `.bp.dir` with sub-files. API-wise both are transparent to the user.

ax3l commented:
@jreadey it seems travis-ci is not enabled by default for the repo. you might want to enable it? (all tests are passing.)

ax3l commented:
@giumas it seems travis-ci is not enabled by default for the repo. you might want to enable it? (all tests are passing.)

jreadey commented:
I've enabled trade-cl for compass.  @ax3l do you want to push something to the branch and verify that that triggers a test run?

ax3l commented:
it seems `adios` is not included by default by the packager yet (`pyinstaller`). do you have any hints what to add? ~~Locally, I just shuffled `adios` into the "hidden imports" to force it to be included, but there might be a clean way (e.g., for a follow-up PR to build the travis aws artefacts correctly).~~ (_did not work_)

ax3l commented:
@jreadey will do!

jreadey commented:
What was the change you made locally?

I'll take a look later today.

ax3l commented:
I tried hacking

``` diff
diff --git a/HDFCompass.1file.spec b/HDFCompass.1file.spec
index d20fd20..f5e1fc6 100644
--- a/HDFCompass.1file.spec
+++ b/HDFCompass.1file.spec
@@ -50,7 +50,8 @@ def collect_pkg_data(package, include_py_files=False, subdir=None):
     return data_toc

 pkg_data_hdf_compass = collect_pkg_data('hdf_compass')
-pkg_data_bag = collect_pkg_data('hydroffice.bag')
+pkg_data_adios = collect_pkg_data('adios')
+#pkg_data_bag = collect_pkg_data('hydroffice.bag')
 cartopy_aux = []
 try:  # for GeoArray we use cartopy that can be challenging to freeze on OSX to dependencies (i.e. geos)
     import cartopy.crs as ccrs
@@ -71,7 +72,7 @@ app_name = 'HDFCompass_' + version
 a = Analysis(['HDFCompass.py'],
              pathex=[],
              hiddenimports=['scipy.linalg.cython_blas', 'scipy.linalg.cython_lapack',
-               'scipy.linalg', 'scipy.integrate'],  # for cartopy
+               'scipy.linalg', 'scipy.integrate', 'adios'],  # for cartopy
              excludes=["PySide"],  # exclude libraries from being bundled (in case that are installed)
              hookspath=None,
              runtime_hooks=None)
@@ -83,7 +84,8 @@ exe = EXE(pyz,
           a.zipfiles,
           a.datas,
           pkg_data_hdf_compass,
-          pkg_data_bag,
+          pkg_data_adios,
+#          pkg_data_bag,
           cartopy_aux,
           name=app_name,
           debug=True,
```

but that did _not_ do the trick as I just found out.

ax3l committed:
License Docs: Add ADIOS

ax3l committed:
Travis: Wait for OSX MPI Install

Avoid timeouts, cache brew builds.

ax3l commented:
ok we can do packing in a follow-up PR. I think it's both necessary to fix the `a = Analysis(['HDFCompass.py'],` line to include the `hdf_compass/compass_viewer/viewer.py` that includes all models and additionally, there seems to be a smaller upstream issue (`_hl` helpers) within adios that I can report/fix first before packing.

giumas commented:
> I tried hacking

This is not the way that I would follow to freeze HDF Compass with the `adios` plugin. 

Is the `adios` package in pure-Python? If not, the issue is most likely that there is not currently a hook for this package in `PyInstaller`: https://github.com/pyinstaller/pyinstaller/tree/develop/PyInstaller/hooks

If you are not familiar, a `PyInstaller`'s hook is a file used to define which additional files should be included by `PyInstaller` (compiled binaries, xml files, images, etc) in order to use a package in a frozen application.

This has nothing to do directly with HDF Compass, since any application using `adios` would have the same issue when freezing using `PyInstaller`. Try to make it working with a minimal file that just import `adios`. The `PyInstaller` developers are very responsive in adding new hooks if you make a PR for it.

giumas commented:
> ok we can do packing in a follow-up PR.

Since the freezing step is very relevant for HDF Compass, I would suggest to first try to fix the issue.

> I think it's both necessary to fix the a = Analysis(['HDFCompass.py'], line to include the viewer.py that includes all models and additionally

I doubt that this is required.

> there seems to be a smaller upstream issue (_hl helpers) within adios that I can report/fix first before packing

Can you post more info on this issue?

ax3l commented:
@giumas thanks for your suggestions.

The adios module is Cython afaik, here is its source: [ornladios/ADIOS](https://github.com/ornladios/ADIOS/tree/master/wrappers/numpy)

I solved it now by this simple diff:

``` diff
diff --git a/HDFCompass.1file.spec b/HDFCompass.1file.spec
index d20fd20..1d5967c 100644
--- a/HDFCompass.1file.spec
+++ b/HDFCompass.1file.spec
@@ -71,7 +71,7 @@ app_name = 'HDFCompass_' + version
 a = Analysis(['HDFCompass.py'],
              pathex=[],
              hiddenimports=['scipy.linalg.cython_blas', 'scipy.linalg.cython_lapack',
-               'scipy.linalg', 'scipy.integrate'],  # for cartopy
+               'scipy.linalg', 'scipy.integrate', 'adios._hl.selections'],  # for cartopy
              excludes=["PySide"],  # exclude libraries from being bundled (in case that are installed)
              hookspath=None,
              runtime_hooks=None)
```

I probably just screwed up my env in the first tries.

The patched line is a hidden include in adios that is otherwise not found ~~at runtime~~. ~~I think I can report that to be fixed upstream instead. depends on your preferences.~~ (_update:_ tried fixing it with renaming without the `_` prefix but did not resolve the need for the hint. I will write the pyinstaller hook for it.)

system info: I am packing on Debian 8.5 "jessie" via `pyinstaller --clean -y HDFCompass.1file.spec`.

ax3l commented:
ok the `PyInstaller` hook PR is open.

Can you try if the OSX (py2app) packing and Windows packing is working? (the ADIOS c-lib and numpy wrapper will likely only be available on linux/osx)

ax3l committed:
ADIOS: Require latest release 1.9.1b19+

jreadey commented:
Is ADIOS supported on Windows?  I was looking at the INSTALL file on github and it mentions Linux and Mac OS X but not our favorite Redmond software vendor.

ax3l commented:
afaik not, it will be a Linux/OSX plugin only. I just wanted to make sure I am not breaking the Windows packaging due to side effects (likely still ok). Maybe you want to enable the appveyor tests on this repo, too?

ax3l commented:
@giumas @jreadey since all tests pass and packaging only depends on the upcoming pyinstaller update (until then: simply not included but other plugins still working) this PR is actually feature complete.

giumas commented on setup.py:
```diff
@@ -105,6 +105,7 @@ def txt_read(*paths):
         "GeoNodes": ["cartopy[plotting]", ],  # required for visualization of GeoArray and GeoSurface nodes
         "BAG": ["hydroffice.bag>=0.2.10", ],  # required by BAG plugin
         "OpenDAP": ["pydap", ],  # required by OpenDAP plugin
+        "ADIOS": ["adios>=1.9.1b19", ],  # required by ADIOS plugin
```
Since the ADIOS plugin is not cross-platform, this should be modified in some how.

ax3l replied:
it's in the `extras_require` section, that should be ok, no?

giumas replied:
**On Windows**, I believe that it would be better that `pip install hdf-compass[adios]` gives back the warning: `hdf-compass does not provide the extra 'adios'` rather than failing in the following way:

```
Collecting adios>=1.9.1b19 (from hdf-compass==0.6.1b1)
  Downloading adios-1.9.1b19.tar.gz (199kB)
    100% |################################| 204kB 2.4MB/s
    Complete output from command python setup.py egg_info:
    adios_config is not installed nor found. Please install Adios or check PATH.
```

ax3l replied:
sounds good, do you know how to code that?

The error message you posted is the default error message if the c-library adios is not installed. The pip (cython) install of the python package `adios` requires the `c library` adios to be already present (as many python packages do). So it is actually a quite clean error message and will work as soon as someone compiled the c library adios on that system, too.

giumas replied:
> sounds good, do you know how to code that?

I am not sure that I understood your question.

You can check if the setup is running on Windows (as we do here: https://github.com/HDFGroup/hdf-compass/blob/develop/hdf_compass/utils/utils.py#L24 ), then add an `if` statement that only adds the `ADIOS` item to the extra requirements dictionary if NOT on Windows.

ax3l replied:
but the thing is: the extra is behaving exactly as it should: is there an adios c library, it can install the extra, otherwise not. independent of the operating system. the pip package is not shipping external dependencies (one of the reasons conda was created.)

that's exactly the same behaviour as trying to install wxPython via pip with missing gstreamer c-library dependencies.

giumas replied:
Right, but it is not clearly defined that ADIOS is not smoothly working on Windows. I understand that there is a way of building it on Windows, but it is not main stream. 

In other words, I don't want that people try to build HDF Compass with all the plugins on Windows, getting an error, have to google around about solutions, and finally finding that GitHub ticket where we discussing about the issue.

If you can give me an example of frozen HDF Compass that I can run on my Windows testing machines, I am fine with leaving as is. Otherwise, I would modify the setup.

giumas replied:
Actually, I have also asked about a frozen HDF Compass for Mac.. Have you been able to freeze the app on Mac?

ax3l replied:
> In other words, I don't want that people try to build HDF Compass with all the plugins on Windows, getting an error, have to google around about solutions, and finally finding that GitHub ticket where we discussing about the issue.

I totally understand that. This is why the main file of this repository, [`README.md` reads](https://github.com/HDFGroup/hdf-compass#development-environment):
- ADIOS Plugin (Linux/OSX only)

> I understand that there is a way of building it on Windows, but it is not main stream. 

and as soon as one can build it, it will just work and this repo needs no update.

> If you can give me an example of frozen HDF Compass that I can run on my testing machines, I am fine with leaving as is. Otherwise, I would modify the setup.

Let us move the freezing questions to the [discussion here](https://github.com/HDFGroup/hdf-compass/issues/155#issuecomment-235557366). Currently, you can just _not install_ the adios python wrapper and nothing will happen beside a "ADIOS Plugin not loaded" terminal message.

giumas commented:
> afaik not, it will be a Linux/OSX plugin only. 

The `README.rst` should be modified to cite it, as well as added the `adios` dependency.

> I just wanted to make sure I am not breaking the Windows packaging due to side effects (likely still ok). Maybe you want to enable the appveyor tests on this repo, too?

I tested on Windows, and the application works showing the message `ADIOS plugin: NOT loaded` as expected.

As a general comment, I am a bit surprised that being a Linux/OSX-only plugin was not specified from the beginning. However, I ran through the code and it looks quite consistent. I leave to @jreadey the burden to accept a plugin that is not cross-platform (all the other plugins are).

Incidentally, @michaelsippel  and @ax3l : can you share a frozen OSX HDF Compass application so that I can test if it works on a clean machine?

jreadey commented:
I think it's ok to have plugins that are platform specific as long as it doesn't introduce problems for supporting the other problems.

@giumas - where you able to build the ADIOS package on linux or mac?   I see it requires MPI, which seems a little odd for a desktop app!

giumas commented:
@jreadey I don't have currently access to my work OSX machine, but I have clean OSX VMs. This is why I asked about the frozen app.

jreadey commented:
@ax3l - do I actually need MPI to build ADIOS?  E.g. I don't have mpicc on my desktop.

What MPI-sh things will the ADIOS Compass plugin do?

ax3l commented:
> The README.rst should be modified to cite it, as well as added the adios dependency.

will do, thanks!

> Windows support

Since ADIOS is a research, HPC-only transport and I/O library by Oak Ridge National Lab this has not been in focus. I am glad we can still use it on the other two platforms, thanks! In the future, the maintainers of ADIOS are actually looking forward to try supporting Windows as well, the progress of the discussion can be followed at https://github.com/ornladios/ADIOS/issues/77 .

> Incidentally, @michaelsippel and @ax3l : can you share a frozen OSX HDF Compass application so that I can test if it works on a clean machine?

yes, I already did build at Linux 64bit ELF a few days ago. You can download it from:
- https://owncloud.hzdr.de/index.php/s/6CSZ5mzxrlUF9d1
- password: compass

> MPI dependency

This is an excellent question and let me shine some light on that. Of course, we do not need any MPI functionality at all and the final plugin also has no MPI dependencies. The only reason we need MPI right now is issue https://github.com/ornladios/ADIOS/issues/74 which hopefully going to be fixed in the mid term, thanks to the maintainer and core developer of ADIOS @pnorbert.

The c-library ADIOS creates two libraries: `libadios` and `libadios_nompi`. The numpy binding then again consists of two independent builds: `adios_mpi` and `adios` (no MPI). We of course only depend on `libadios_nompi` and `adios`. The only problem is currently, that ADIOS' build scripts can only build the two targets at once. I hope that is no problem for now and I can reduce the travis build scripts as soon as https://github.com/ornladios/ADIOS/issues/74 is fixed (I also linked a docker script there for building it).

ax3l committed:
Readme: Add ADIOS Plugin

Add documentation & links to ADIOS plugin.
Update packing (same on Win as on Linux).

giumas commented:
> yes, I already did build at Linux 64bit ELF a few days ago.

@ax3l : since I asked for a Mac OSX frozen app,  what do you mean with that?

On the same topic, if you were successful in freezing the app on OSX: 
- Which environment did you use to freeze the app? 
- Which Mac OSX flavor? 

We need this information since they will be useful when it will come the time of a new release. This also helps @jreadey to track the minimum supported/tested OS. In particular, this is very relevant for Mac OSX: https://pythonhosted.org/PyInstaller/usage.html#making-mac-os-x-apps-forward-compatible 

In other words, we ideally want to achieve a statement like this: "A working frozen HDF Compass app with the `adios` plugin has been built and tested on: 
- OSX [10.8, 10.9, 10.10, 10.11]
- [Centos x.x]
- [Ubuntu x.x]
- [...]
  ". 

If for any reason we are not currently able to freeze the app on OSX we might start by making the plugin Linux-only.

@ghisvail : what are your Debianic thoughts about this optional plugin?

ax3l commented:
> > > Incidentally, @michaelsippel and @ax3l : can you share a frozen OSX HDF Compass application so that I can test if it works on a clean machine?
> > 
> >  yes, I already did build at Linux 64bit ELF a few days ago.
> 
> @ax3l : since I asked for a Mac OSX frozen app, what do you mean with that?

@giumas sorry for the misunderstanding, ~~you mentioned your OSX VMs in a later post~~ I oversaw the "OSX" before "HDF". I had no hands on OSX yet, that's why I am asking if someone can try the build there. As I said, travis-builds on OSX 10.9 work as you can see [here](https://travis-ci.org/HDFGroup/hdf-compass/jobs/144697798) - but these scripts do not try packing it.

I like the way to list and test the supported versions. So far I did the HDFCompass tests on:
- building on travis trusty works
- building and packing on Debian 8.5 jessie works
- testing the packed linux ELF works on latest Debian 8.5 "jessie", Mint 17.3 "rosa", Ubuntu 14.04 "trusty"  and 16.4 "xenial" , etc. (could not spot problems with those for everyone I asked to try running it)

Building adios (c lib and numpy wrapper) worked for me on any *nix I came around:
- Debian latest (c lib also in deb [packages](https://packages.debian.org/jessie/libadios-dev))
- Red hat 6.7 (Rhea/ORNL)
- Cray Linux XK7 (Titan/ORNL)
- Ubuntu 12+ (c lib: [packages](http://packages.ubuntu.com/trusty/libadios-dev))
- Ubuntu 14.04 (Hypnos/HZDR)
- Scientific Linux 7.6 (Lawrencium/LBNL)
- Ubuntu "xenial" (docker)
- arch linux (aur packet [adios](https://aur.archlinux.org/packages/adios/))

Any OSX testing is super-welcome, I have no VMs around for it but know the ADIOS developers use OSX machines and @yyalli and @suchyta1 are currently testing it out (feel free to join with your test results!).

Regarding the `pyinstaller` text you linked: aren't you using `py2app` for OSX packing? (Your readme states so.)

ax3l committed:
Travis: C-ADIOS without MPI

ax3l commented:
Updated the following aspects:
- ADIOS now builds `--without-mpi` (less build-time dependencies)
- ~~more caching to reduce traffic & build times on travis~~ (_out of scope for this PR_)

jreadey commented:
I've merged to the develop branch!

ax3l commented:
thank you! :fireworks:
